File: C/data_structures/stack/stack.c
// Stack implementation in C

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

#define MAX 1000
int stack[MAX];

// Initialize top of stack
int top = -1;

// Function to add an item to stack
void push(int data) {
  if (top == MAX - 1) {
    printf("Stack Overflow\n");
    return;
  }
  stack[++top] = data;
}

// Function to remove an item from stack
int pop() {
  if (top == -1) {
    printf("Stack Underflow\n");
    return INT_MIN;
  }
  return stack[top--];
}

// Function to return the top from stack without removing it
int peek() {
  if (top == -1) {
    printf("Stack is empty\n");
    return INT_MIN;
  }
  return stack[top];
}

// helper functions
int isEmpty() { return top == -1; }
int isFull() { return top == MAX - 1; }
int size() { return top + 1; }

// Function to display stack elements
void display() {
  if (top == -1) {
    printf("Stack is empty\n");
    return;
  }
  for (int i = top; i >= 0; i--) {
    printf("%d ", stack[i]);
  }
  printf("\n");
}

/**
 * Main function to demonstrate stack operations
*/
int main() {
  push(10); // Push 10 onto the stack
  push(20); // Push 20 onto the stack
  push(30); // Push 30 onto the stack
  display(); // Display stack elements
  printf("Top element is %d\n", peek()); // Peek at the top element
  printf("Stack size is %d\n", size()); // Get the size of the stack
  printf("Popped element is %d\n", pop()); // Pop the top element
  display(); // Display stack elements after pop
  return 0;
}

File: JavaScript/data_structures/stack/stack.js
class Stack {
  constructor() {
    this.items = [];
  }

  push(element) {
    this.items.push(element);
  }

  pop() {
    if (this.isEmpty()) {
      return "Stack is empty";
    }
    return this.items.pop();
  }

  peek() {
    if (this.isEmpty()) {
      return "Stack is empty";
    }
    return this.items[this.size() - 1];
  }

  isEmpty() {
    return this.size() === 0;
  }

  size() {
    return this.items.length;
  }

  display() {
    return `[${this.items.join(", ")}]`;
  }

  clear() {
    this.items = [];
  }
}

// Test cases
function testStack() {
  console.log("---------- Stack Test Cases ----------");

  const stack = new Stack();

  console.log("Test 1 - isEmpty():", stack.isEmpty());

  stack.push(10);
  stack.push(20);
  stack.push(30);

  console.log("Test 2 - After pushing 10, 20, 30:", stack.display());

  console.log("Test 3 - Size:", stack.size());

  console.log("Test 4 - Peek:", stack.peek());

  console.log("Test 5 - Pop:", stack.pop());

  console.log("Test 5 - After pop:", stack.display());

  console.log("Test 6 - isEmpty() after operations:", stack.isEmpty());

  stack.pop();
  stack.pop();

  console.log("Test 7 - After popping all:", stack.display());

  console.log("Test 8 - Pop from empty:", stack.pop());

  console.log("Test 9 - Peek empty:", stack.peek());

  stack.push(1);
  stack.push(2);

  stack.clear();

  console.log("Test 10 - After clear:", stack.isEmpty());
}

testStack();